[ONPREM-1092] Add machine runner 3.0 docker docs (#8729)
circleci/circleci-docs: jekyll/_cci2/install-machine-runner-3-on-docker.adoc

---
contentTags:
  platform:
  - Cloud
  - Server v4.4+
---
= Install machine runner 3.0 on Docker
:page-layout: classic-docs
:page-liquid:
:page-description: Instructions on how to install CircleCI's self-hosted machine runner (3.0) on Docker.
:icons: font
:experimental:
:machine:
:docker:

> This page describes how to install CircleCI's machine runner 3.0 with the Docker executor on server. If you are looking to set up self-hosted runners in a private Kubernetes cluster, visit the <<container-runner-installation#,Container runner>> page.

Reviewer: ```suggestion
This page describes how to install CircleCI's machine runner 3.0 with the Docker executor. If you are looking to set up self-hosted runners in a private Kubernetes cluster, visit the <<container-runner-installation#,Container runner>> page.
```

[#new-recommended-method-container-runner]

> == Recommended method for Docker installation (container runner)

Reviewer: Is this section to tell people that following this guide is NOT the recommended method?

Author: Yes exactly. This is really only a backup for customers who want to use a simple docker host setup rather than replicating cloud docker jobs on runner

The <<container-runner#,container runner>> is the recommended approach for running containerized jobs on self-hosted runners. Container runner offers the ability to seamlessly define, publish, and use custom Docker images during job execution, as well as the ability to manage dependencies or libraries through custom Docker images instead of enumerating dependencies as part of `steps` in the `.circleci/config.yml` file.

[#machine-approach-with-docker]
== Machine-based approach with Docker

[#machine-runner-prerequsites]
=== Prerequisites

* link:https://docs.docker.com/engine/install/[Docker]

> [#additional-prerequisites]
> ==== Additional prerequisites with Docker

Reviewer: ```suggestion
[#resource-requirements]
==== Resource requirements
```

Reviewer: Does this title make sense?

Author: Yes that looks good

{% include snippets/runner/terms.adoc %}

[#create-namespace-and-resource-class]
=== 1. Create namespace and resource class

{% include snippets/runner/install-with-cli-steps.adoc %}

=== 2. Create a Dockerfile that extends the machine runner 3.0 image

Create a `Dockerfile.runner.extended` file. In this example, Python 3 is installed on top of the base image.

```dockerfile
FROM circleci/runner-agent:machine-3
RUN sudo apt-get update; \
    sudo apt-get install --no-install-recommends -y \
        python3
```

[#build-the-docker-image]
=== 3. Build the Docker image

```shell
docker build --file ./Dockerfile.runner.extended .
```

[#start-the-docker-container]
=== 4. Start the Docker container

NOTE: The environment variable values are not available to the `docker` command, so these environment variables are not visible in `ps` output.

```shell
CIRCLECI_RUNNER_NAME=<runner-name> CIRCLECI_RUNNER_API_AUTH_TOKEN=<runner-token> docker run --env CIRCLECI_RUNNER_NAME --env CIRCLECI_RUNNER_API_AUTH_TOKEN --name <container-name> <image-id-from-previous-step>
```

When the container starts, it will immediately attempt to start running jobs.

[#start-the-docker-container-on-server]

> ==== Start the Docker container on server

Reviewer: I'll put this part in Cloud/Server tabs in a separate commit

> When starting the Docker container on server, the `agent_version` and `LAUNCH_AGENT_API_URL` environment variables will need to be passed in using the `--env` flag.

Reviewer 2: s/LAUNCH_AGENT_API_URL/CIRCLECI_RUNNER_API_URL

Reviewer 2: Maybe a hold over from the launch-agent docs? I don't think we specify an agent version any more, or at least I don't see it in the command options when calling help on the machine mode.

```shell
CIRCLECI_RUNNER_NAME=<runner-name> CIRCLECI_RUNNER_API_AUTH_TOKEN=<runner-token> CIRCLECI_RUNNER_API_URL=<server_host_name> docker run --env CIRCLECI_RUNNER_NAME --env CIRCLECI_RUNNER_API_AUTH_TOKEN --env CIRCLECI_RUNNER_API_URL --name <container-name> <image-id-from-previous-step>
```

[#stopping-the-docker-container]
=== Stopping the Docker container

```shell
docker stop <container-name>
```

[#remove-the-docker-container]
=== Remove the Docker container

In some cases you might need to fully remove a stopped machine runner container from the system, such as when recreating a container using the same name.

```shell
docker stop <container-name>; docker rm <container-name>;
```

[#migrating-from-launch-agent]
=== Migrating from launch agent

> To migrate from launch agent to machine runner 3.0 on Docker simply requires stopping & removing the launch agent based containers and replacing them with the machine runner 3.0 based containers, as described above. The prefix of the name of environment variables has changed from `LAUNCH_AGENT_` to `CIRCLECI_RUNNER_`.

Reviewer: ```suggestion
NOTE: In machine runner 3.0, the environment variable name prefix has changed from `LAUNCH_AGENT_` to `CIRCLECI_RUNNER_`.

To migrate from launch agent to machine runner 3.0 on Docker, stop and remove the launch agent containers and replace them with machine runner 3.0 containers, using the commands described above. 
```

Reviewer: Might be better to rework this as a step by step process but I will do that in a separate commit

---
contentTags:
  platform:
  - Cloud
  - Server v4.4+
---
= Install machine runner 3.0 on Docker
:page-layout: classic-docs
:page-liquid:
:page-description: Instructions on how to install CircleCI's self-hosted machine runner (3.0) on Docker.
:icons: font
:experimental:
:machine:
:docker:

This page describes how to install CircleCI's machine runner 3.0 with the Docker executor. If you are looking to set up self-hosted runners in a private Kubernetes cluster, visit the <<container-runner-installation#,Container runner>> page.

[NOTE]
====

> xref:container-runner#[Container runner] is the **recommended method** for self-hosted runner Docker installation. The instructions on this page are for a simple Docker setup using machine runner 3.0.

Reviewer 2: We had some confusion in a call with horizon yesterday when I was trying to explain that docker runners are machine runners. Could a CI config example like we have for machine runner 3 (https://circleci.com/docs/install-machine-runner-3-on-linux/#machine-runner-configuration-example) be added in here as well so we avoid folks trying to use the docker key?

Author: That is already included in the web app [install snippet](https://internal.circleci.com/docs/runner-installation-docker/#create-namespace-and-resource-class)

I would like to think that plus all the references to it being a machine runner is sufficient?

Author: Ignore that 😅 I have added the snippet used on the page you linked
